### ISSUE TITLE
Check for "constants we match on must be PartialEq" misses lifetime-dependent impls

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#121007]
The following should trigger a warning that we can only match on PartialEq consts, but it doesn't: (example by @lcnr)
```rust
use std::marker::PhantomData;

struct Inv<'a>(PhantomData<*mut &'a ()>);

impl<'a> PartialEq<Inv<'static>> for Inv<'a> {
    fn eq(&self, _: &Inv<'static>) -> bool {
        true
    }
}

impl<'a> Inv<'a> {
    const NOT_STATIC: Option<Self> = None;
}

fn foo<'a>(x: Option<Inv<'a>>) {
    match x {
        Inv::<'a>::NOT_STATIC => (),
        Some(_) => panic!()
    }
    
    //x == Inv::<'a>::NOT_STATIC;
}

fn main() {
    foo(None)
}
```

Here is [another example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=183a1d779e2b26eabec2dc43ee31a40f) of the same issue.

The problem is that we call `predicate_must_hold_modulo_regions` [here](https://github.com/rust-lang/rust/blob/9f58cf43c78e77b59479da2beaea0265f529280e/compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs#L257), but only borrowck knows the actual lifetimes and borrowck has no idea that this trait obligation even exists.

We should leave some sort of trace in MIR that there is a PartialEq obligations to ensure borrowck can check this with the right lifetimes.

**Comment by fmease** [comment]
I guess we could so something similar to #105102?

**Comment by RalfJung** [comment]
Maybe? I have absolutely no idea what that PR does.^^ I don't really know anything about our trait system implementation.

**Comment by lcnr** [comment]
> I guess we could do something similar to #105102?

sadly not, the "type is copy" check can be self-contained and can therefore handle its region constraints by itself using `infcx.resolve_regions`. As patterns are part of MIR bodies, we have to check their region constraints together with the rest of the body during MIR borrowck. The implementation would instead be similar to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/typeck_results/struct.CanonicalUserTypeAnnotation.html and https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.Body.html#structfield.required_consts: we store a list of requirements which have to be true for the given MIR even though they aren't directly present in the source. We then check them during mir borrowck
